### PR TITLE
train mobilenetv3-large

### PR DESCRIPTION
**Pull request by fyrirvari**

### CODE REVIEW

**Comment by kprokofi** [comment]
1) We don't need so much .pth weights
We need only one for mobilenetv3 large. Please, get rid of unnecessary weights
2) we don't need cuda test in our project
3) we don't use .ipnb files. We need runable scripts .py files

Please, make an appropriate changes and I look through again

File: common/mmodel.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    return torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
+
+def save_model(model, path="pretrained"):
+    torch.save(model.state_dict(), os.path.join(path, "model_weights.pth"))
+    torch.save(model, os.path.join(path, "model.pth"))
+
+def load_model(path="pretrained"):
+    model = torch.load(os.path.join(path, "model.pth"))
+    model.load_state_dict(torch.load(os.path.join(path, "model_weights.pth")))
+    model.eval()
```
Comment (kprokofi) [inline common/mmodel.py]: why do we need to set eval mode in the loading function?

File: common/mmodel.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    return model
+
+def create_dataloaders():
+    data_transforms = {
+        'train': transforms.Compose([
+            transforms.RandomResizedCrop(224),
```
Comment (kprokofi) [inline common/mmodel.py]: Don't do a random resize. We have images 28x28. We upscale them and we can't randomly choose the area. Set just a resize augmentation

File: common/mmodel.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    torch.save(model.state_dict(), os.path.join(path, "model_weights.pth"))
+    torch.save(model, os.path.join(path, "model.pth"))
+
+def load_model(path="pretrained"):
+    model = torch.load(os.path.join(path, "model.pth"))
+    model.load_state_dict(torch.load(os.path.join(path, "model_weights.pth")))
+    model.eval()
+    return model
+
+def create_dataloaders():
```
Comment (kprokofi) [inline common/mmodel.py:36]: We do not need this function. Ivan has created a specific one for our case. I will merge his pull request and then you will add his function. From this function, we only need augmentations.

File: common/mmodel.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+                for x in ['train', 'val']}
+
+    return dataloaders
+
+def get_classes(path="data_json/names.csv", sep=","):
+    classes = pd.read_csv(path, sep=sep).iloc[:,1]
+    return classes.tolist()
+
+def pred_images(model, images):
+    img_transforms = transforms.Compose([
```
Comment (kprokofi) [inline common/mmodel.py]: another one transforms?

File: common/pipeline.py [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+
+    # load best model weights
+    model.load_state_dict(best_model_wts)
+    return model
+
+def train_mnetv3(num_epochs=25):
+    # Data augmentation and normalization for training
+    # Just normalization for validation
+    data_transforms = {
+        'train': transforms.Compose([
```
Comment (kprokofi) [inline common/pipeline.py]: third transforms? :)

File: common/pipeline.py [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+    device = mmodel.get_device()
+
+    model_ft = mobilenetv3.mobilenetv3_large(num_classes=len(class_names))
+
+    model_ft = model_ft.to(device)
+
+    criterion = nn.CrossEntropyLoss()
+
+    # Observe that all parameters are being optimized
+    optimizer_ft = optim.SGD(model_ft.parameters(), lr=0.001, momentum=0.9)
```
Comment (kprokofi) [inline common/pipeline.py]: I guess it is better to use an Adam optimizer with a 0.001 LR value or increase the learning rate to 0.013 for SGD. (Choose what you want)

File: common/pipeline.py [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+
+    # Decay LR by a factor of 0.1 every 7 epochs
+    exp_lr_scheduler = lr_scheduler.StepLR(optimizer_ft, step_size=7, gamma=0.1)
```
Comment (kprokofi) [inline common/pipeline.py]: use cosine scheduler. Steps need to be fine-tuned. We don't have time

File: common/pipeline.py [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+
+            print('{} Loss: {:.4f} Acc: {:.4f}'.format(
+                phase, epoch_loss, epoch_acc))
+
+            # deep copy the model
+            if phase == 'val' and epoch_acc > best_acc:
+                best_acc = epoch_acc
+                best_model_wts = copy.deepcopy(model.state_dict())
+
+        print()
```
Comment (kprokofi) [inline common/pipeline.py:70]: empty print?

File: common/pipeline.py [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+import time
+import os
+import copy
+
+from common import mobilenetv3, data_loader, pipeline, mmodel
+
+def train_model(model, dataloaders, criterion, optimizer, scheduler, device, dataset_sizes, num_epochs=25):
+    since = time.time()
+
+    best_model_wts = copy.deepcopy(model.state_dict())
```
Comment (kprokofi) [inline common/pipeline.py]: please substitute the deep copying in favor of saving weights on the local disk

File: requirements.txt [hunk 10]
```diff
@@ -0,0 +1,54 @@
+async-generator==1.10
```
Comment (kprokofi) [inline requirements.txt]: do we really need all of these packages to run your training?

**Review by kprokofi — CHANGES_REQUESTED** [review]

File: common/mmodel.py (hunk 1 above)
Comment (fyrirvari) [reply common/mmodel.py]: i suppose that when we load the model we will use it to predict.
ok, i will move this to a function for prediction

File: common/pipeline.py (hunk 8 above)
Comment (fyrirvari) [reply common/pipeline.py:70]: for more visual output to the console when model is being trained

File: requirements.txt (hunk 10 above)
Comment (fyrirvari) [reply requirements.txt]: Whoops

File: common/pipeline.py (hunk 5 above)
Comment (fyrirvari) [reply common/pipeline.py]: c:

File: common/mmodel.py (hunk 1 above)
Comment (kprokofi) [reply common/mmodel.py]: Yes, it is better to switch this mode in the place of model evaluation

**Review by kprokofi — APPROVED** [review]
Please, change Adam to AdamW when you will train the model
